docs/querying/datasource.md

```diff
@@ -22,43 +22,317 @@ title: "Datasources"
+[Druid SQL](sql.md#query-translation) documentation.
+
+#### Joins in native queries
+
+Native join datasources have the following properties. All are required.
+
+|Field|Description|
+|-----|-----------|
+|`left`|Left-hand datasource. Must be of type `table`, `join`, `lookup`, `query`, or `inline`. Placing another join as the left datasource allows you to join arbitrarily many datasources.|
+|`right`|Right-hand datasource. Must be of type `lookup`, `query`, or `inline`. Note that this is more rigid than what Druid SQL requires.|
+|`rightPrefix`|String prefix that will be applied to all columns from the right-hand datasource, to prevent them from colliding with columns from the left-hand datasource. Can be any string, so long as it is nonempty and is not be a prefix of the string `__time`. Any columns from the left-hand side that start with your `rightPrefix` will be shadowed. It is up to you to provide a prefix that will not shadow any important columns from the left side.|
+|`condition`|[Expression](../misc/math-expr.md) that must be an equality where one side is an expression of the left-hand side, and the other side is a simple column reference to the right-hand side. Note that this is more rigid than what Druid SQL requires: here, the right-hand reference must be a simple column reference; in SQL it can be an expression.|
+|`joinType`|`INNER` or `LEFT`.|
+
+#### Join performance
+
+Joins are a feature that can significantly affect performance of your queries. Some performance tips and notes:
+
+1. Joins are especially useful with [lookup datasources](#lookup), but in most cases, the
+[`LOOKUP` function](sql.html#string-functions) performs better than a join. Consider using the `LOOKUP` function if
+it is appropriate for your use case.
+2. When using joins in Druid SQL, keep in mind that it can generate subqueries that you did not explicitly include in
+your queries. Refer to the [Druid SQL](sql.md#query-translation) documentation for more details about when this happens
+and how to detect it.
+3. One common reason for implicit subquery generation is if the types of the two halves of an equality do not match.
+For example, since lookup keys are always strings, the condition `druid.d JOIN lookup.l ON d.field = l.field` will
+perform best if `d.field` is a string.
+4. As of Druid {{DRUIDVERSION}}, the join operator must evaluate the condition for each row. In the future, we expect
+to implement both early and deferred condition evaluation, which we expect to improve performance considerably for
+common use cases.
+
+#### Future work for joins
+
+Joins are an area of active development in Druid. The following features are missing today but may appear in
+future versions:
+
+- Preloaded dimension tables that are wider than lookups (i.e. supporting more than a single key and single value).
+- RIGHT OUTER and FULL OUTER joins. Currently, they are partially implemented. Queries will run but results will not
```

Comment:
Is it possible to issue these queries today? I tried doing it from SQL and the query wouldn't plan. IIRC native queries also do not allow these conditions.

Maybe something like 
CROSS joins are supported via sql, but it does so using sub-queries, so it really should not be used in production use cases

Reply:
There are some tests for the functionality and they do pass. They're in CalciteQueryTest (SQL) and HashJoinSegmentStorageAdapterTest (native). What query did you try to run?